### PR TITLE
Preserve order of response headers

### PR DESCRIPTION
**Pull request by davidfstr**
Patch for issue: https://github.com/shazow/urllib3/issues/821

### CODE REVIEW

**Comment by davidfstr** [comment]
Apparently still needs some work to function on Python 2.x. Will take another look in a day or two.

File: test/test_connectionpool.py [hunk 1]
```diff
@@ -308,7 +311,80 @@ def kaboom(*args, **kwargs):
...
+            conn = original_new_conn(*args, **kwargs)
+            conn.request = lambda *args, **kwargs: None
+            conn.getresponse = lambda *args, **kwargs: httplib_response
+            return conn
+        c._new_conn = patched_new_conn
+        
+        actual_response = c.urlopen('GET', '/', release_conn=False)
+        self.assertEqual(
+            list(actual_response.headers.items()),
+            list(expected_response_headers.items()))
```
Comment (Lukasa) [inline test/test_connectionpool.py]: These tests almost certainly require more work than simply writing a socketlevel test like we have elsewhere. You may want to give that a shot.

**Comment by davidfstr** [comment]
The last failing test is the same one that is failing on the `master` branch:

* ERROR: test_delayed_body_read_timeout (test.with_dummyserver.test_socketlevel.TestSocketClosing)

I'll wait until `master` is passing, then I'll rebase & squish this branch to be more readable.

File: test/test_connectionpool.py (hunk 1 above)
Comment (davidfstr) [inline test/test_connectionpool.py]: I didn't understand your suggestion. Could you rephrase?

File: test/test_connectionpool.py [hunk 2]
```diff
@@ -308,7 +311,91 @@ def kaboom(*args, **kwargs):
...
+    
+    def test_preserves_request_header_order(self):
+        """
+        Ensure that headers provided to ConnectionPool.urlopen()
+        are sent to the underlying httplib in the same order.
+        """
+        class AbortRequest(BaseException):
+            pass
+        
+        expected_request_headers = OrderedDict([('X-Header-%d' % i, str(i)) for i in range(16)])
```
Comment (sigmavirus24) [inline test/test_connectionpool.py]: Why are these tests making 16 headers? Why not 5? Why not 20? What is special about 16?

File: test/test_connectionpool.py (hunk 1 above)
Comment (sigmavirus24) [inline test/test_connectionpool.py]: @davidfstr you're duplicating a lot of code between these tests when you could instead write a different type of test (a socket-level test) with the dummyserver that would require a lot less work in each test, be easier to read, understand, and maintain.

Comment (davidfstr) [inline test/test_connectionpool.py]: Sounds good to me. I'll take a look.

File: test/test_connectionpool.py (hunk 2 above)
Comment (davidfstr) [inline test/test_connectionpool.py]: Nothing too special. Just a large enough number that it's a low probability that (shuffle(X) == X), where X is a list of headers. Maybe I'll change it to something like 8 just so that it's not a power of 2 (which might imply the number was special). :-)

Comment (Lukasa) [inline test/test_connectionpool.py]: Eh, I think I'd go the other way: make it vanishingly unlikely that the test passes. 500 entries, say.

Comment (shazow) [inline test/test_connectionpool.py]: Right now it gets inserted in alphabetical order, no? How about we just random.shuffle(...) it and check that we get it back in the same shuffled order? That should be fine with a small number of entries.

**Comment by davidfstr** [comment]
Rescinding this PR in favor of: https://github.com/shazow/urllib3/pull/827
